chalicelib/checks/wrangler_checks.py

```diff
@@ -2032,3 +2032,93 @@ def add_contributing_lab_opf(connection, **kwargs):
         action.status = 'DONE'
     action.output = action_logs
     return action
+
+
+@check_function()
+def grouped_with_file_relation_consistency(connection, **kwargs):
+    ''' Check if "grouped with" file relationships are reciprocal and complete.
+        While other types of file relationships are automatically updated on
+        the related file, "grouped with" ones need to be explicitly (manually)
+        patched on the related file. This check ensures that there are no
+        related files that lack the reciprocal relationship, or that lack some
+        of the group relationships (for groups larger than 2 files).
+    '''
+    check = CheckResult(connection, 'grouped_with_file_relation_consistency')
+    check.action = 'add_grouped_with_file_relation'
+    search = 'search/?type=File&related_files.relationship_type=grouped+with&field=related_files'
+    files = ff_utils.search_metadata(search, key=connection.ff_keys, is_generator=True)
+
+    file2all = {}  # map all existing relations
+    file2grp = {}  # map "group with" existing relations
+    for f in files:
+        for rel in f['related_files']:
+            rel_type = rel['relationship_type']
+            rel_file = rel['file']['@id']
+            file2all.setdefault(f['@id'], []).append(
+                {"relationship_type": rel_type, "file": rel_file})
+            if rel_type == "grouped with":
+                file2grp.setdefault(f['@id'], []).append(rel_file)
+
+    # list groups of related items
+    groups = []
+    newgroups = [set(rel).union({file}) for file, rel in file2grp.items()]
+
+    # merge intersecting groups until there is any
```

Comment:
Not sure what you mean exactly by this comment.

Reply:
Given the pairwise nature of file relationships, groups can be partially defined on each file. 
An initial list of (partial) groups is compiled by reading the relationships of each file. Now, if two partial groups have one or more common items, they need to be merged into one group. This needs to be repeated recursively on the list of partial groups, until there is no intersection.